lib/rexml/parsers/baseparser.rb

```diff
@@ -635,21 +635,25 @@ def process_instruction(start_position)
           @source.position = start_position
           raise REXML::ParseException.new(message, @source)
         end
-        if @document_status.nil? and match_data[1] == "xml"
-          content = match_data[2]
-          version = VERSION.match(content)
-          version = version[1] unless version.nil?
-          encoding = ENCODING.match(content)
-          encoding = encoding[1] unless encoding.nil?
-          if need_source_encoding_update?(encoding)
-            @source.encoding = encoding
-          end
-          if encoding.nil? and /\AUTF-16(?:BE|LE)\z/i =~ @source.encoding
-            encoding = "UTF-16"
+        if match_data[1] == "xml"
+          if @document_status
+            raise ParseException.new("Malformed XML: XML declaration other than at the top of the document.", @source)
+          else
```

Comment:
* How about early return style to reduce indent?
* Could you remove `.` in the error message?
* Can we simplify the error message?

```suggestion
          unless @document_status
            raise ParseException.new("Malformed XML: XML declaration is not at the start", @source)
          end
```

Reply:
`unless @document_status` would result in the opposite branch.
The correct answer is `if @document_status`, right?

```
          if @document_status
            raise ParseException.new("Malformed XML: XML declaration is not at the start", @source)
          end
```

Reply:
Oh, sorry.